moveit_ros/benchmarks/src/BenchmarkExecutor.cpp

```diff
@@ -897,6 +922,111 @@ void BenchmarkExecutor::collectMetrics(PlannerRunData& metrics,
   }
 }
 
+void BenchmarkExecutor::computeResultPathSimilarity(
+    PlannerBenchmarkData& planner_data, const std::vector<planning_interface::MotionPlanDetailedResponse>& mp_res,
+    const std::vector<bool>& solved)
+{
+  ROS_INFO("Computing result path similarity");
+  const size_t result_count = planner_data.size();
+  size_t unsolved = std::count_if(solved.begin(), solved.end(), [](bool s) { return !s; });
+  std::vector<double> average_distances(mp_res.size());
+  for (size_t first_traj_i = 0; first_traj_i < result_count; ++first_traj_i)
+  {
+    // If trajectory was not solved there is no valid average distance so it's set to max double only
+    if (!solved[first_traj_i])
+    {
+      average_distances[first_traj_i] = std::numeric_limits<double>::max();
+      continue;
+    }
+    // Iterate all result trajectories that haven't been compared yet
+    for (size_t second_traj_i = first_traj_i + 1; second_traj_i < result_count; ++second_traj_i)
+    {
+      // Ignore if other result has not been solved
+      if (!solved[second_traj_i])
+        continue;
+
+      // Abort if there are no result trajectories
+      if (mp_res[first_traj_i].trajectory_.empty() || mp_res[second_traj_i].trajectory_.empty())
+        continue;
+
+      // Access trajectories
+      const robot_trajectory::RobotTrajectory& traj_first = *mp_res[first_traj_i].trajectory_.back();
+      const robot_trajectory::RobotTrajectory& traj_second = *mp_res[second_traj_i].trajectory_.back();
+
+      // Abort if trajectories are empty
+      if (traj_first.empty() || traj_second.empty())
+        continue;
+
+      // Waypoint counter
+      size_t pos_first = 0;
+      size_t pos_second = 0;
+      const size_t max_pos_first = traj_first.getWayPointCount() - 1;
+      const size_t max_pos_second = traj_second.getWayPointCount() - 1;
+
+      // Compute total distance between pairwise waypoints of both trajectories.
+      // The selection of waypoint pairs is based on what steps results in the minimal distance between the next pair of
+      // waypoints. We first check what steps are still possible or if we reached the end of the trajectories. Then we
+      // compute the pairwise waypoint distances of the pairs from increasing both, the first, or the second trajectory.
+      // Finally we select the pair that results in the minimal distance, summarize the total distance and iterate
+      // accordingly. After that we compute the average trajectory distance by normalizing over the number of steps.
+      double total_distance = 0;
+      size_t steps = 0;
+      double current_distance = traj_first.getWayPoint(pos_first).distance(traj_second.getWayPoint(pos_second));
+      while (true)
+      {
+        // Keep track of total distance and number of comparisons
+        total_distance += current_distance;
```

Comment:
Is this just [DynamicTimeWarping](https://ompl.kavrakilab.org/DynamicTimeWarp_8cpp_source.html#l00058)? The idea looks really similar, the only difference I can see is that at the end of this you're dividing by `steps`, which could be different than `max(max_pos_first, max_pos_second)`, since it's the actual number of entries in the table being summed. If it is supposed to be the same, should we just call it that?

@davetcoleman, you wrote that implementation of Dynamic Time Warping, right?

Reply:
It has some similarities but DynamicTimeWarp minimizes the distance over all waypoint pairs O(n*m) while this algorithm attempts to only compare only waypoints that should correspond to the same position in the trajectories O(n+m). Of course this approach is really just a heuristic that only works for similar trajectories but from my experiments planners can be compared reliably with this metric.

Reply:
Yes, I wrote that long ago :-)